### PR TITLE
EEJT-0002

### PR DESCRIPTION
**Pull request by treep25**

### CODE REVIEW

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+@RequestMapping(value = "/certificates", produces = MediaType.APPLICATION_JSON_VALUE)
+public class GiftCertificateController {
+    private final GiftCertificateService giftCertificateService;
+
+    @Autowired
+    public GiftCertificateController(GiftCertificateService giftCertificateService) {
+        this.giftCertificateService = giftCertificateService;
+    }
+
+    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE)
```
Comment (VladyslavBieliaiev) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java:25]: Here we can use just PostMapping,  becase RestController by default set JSON types

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    public ResponseEntity<?> createCertificate(@RequestBody GiftCertificate giftCertificate) {
+
+        if (DataValidation.isValidCertificate(giftCertificate)) {
+            giftCertificateService.createCertificate(giftCertificate);
+
+            return new ResponseEntity<>(Map.of("status", HttpStatus.CREATED), HttpStatus.CREATED);
+        }
+        throw new IllegalArgumentException("Something went wrong during the request, check your fields");
```
Comment (VladyslavBieliaiev) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java:33]: in such case, it's better to throw specifix exception or rreturn ResponseEntity here or add validation annotations to the GiftCertificate

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @GetMapping()
```
Comment (VladyslavBieliaiev) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java:36]: We could use just `@GetMapping`

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        this.tagService = tagService;
+    }
+
+
+    public List<GiftCertificate> getAllCertificates() {
+        List<GiftCertificate> giftCertificateList = giftCertificateRepository.getAllGiftCertificates();
+        if (!giftCertificateList.isEmpty()) {
+            return giftCertificateList;
+        }
+        throw new ItemNotFoundException("There are no gift certificates");
```
Comment (VladyslavBieliaiev) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java:34]: For search, get all we can return empty list as a response

File: EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import java.sql.SQLSyntaxErrorException;
+import java.util.Map;
+
+@ControllerAdvice
+public class ResponseExceptionHandler extends ResponseEntityExceptionHandler {
+
+    @ExceptionHandler(value = {IllegalArgumentException.class, IllegalStateException.class})
+    protected ResponseEntity<?> handleConflictException(RuntimeException ex, WebRequest request) {
+        return handleExceptionInternal(ex,
+                Map.of("HTTP Status", HttpStatus.CONFLICT, "response body", Map.of("error message", ex.getMessage())),
```
Comment (VladyslavBieliaiev) [inline EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java:21]: you throw IllegalArgumentException when client send request with invalid data, its' not the 409 CONFLIT status, it's 400. Server may return 409, when client try to update resource which have the newest version than in client request or for example specific business logic for tickets to cinema, train

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java (hunk 1 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java:25]: fix it

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java (hunk 2 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java:33]: I use specific ServerException ; fix it

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java (hunk 3 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java:36]: fix it

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java (hunk 4 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java:34]: fix it , i hope in getById it is not useful to return empty list , I return list or else return ItemNotFountEx

File: EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java (hunk 5 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java:21]: fix it

**Review by OlehHoliuk — CHANGES_REQUESTED** [review]

File: EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/repository/GiftCertificateTagRepository.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import com.epam.esm.giftcertficate.model.GiftCertificate;
+
+import java.util.List;
+
+public interface GiftCertificateTagRepository {
+    List<GiftCertificate> getGiftCertificateTagsByTagName(String tagName);
+
+    List<GiftCertificate> getGiftCertificatesAndTagsByNameOrByPartOfName(String partOfName);
+
+    List<GiftCertificate> sortingAscDescByDate(String method);
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/repository/GiftCertificateTagRepository.java:12]: we need sort byDate asc/desc , i don`t understand this (

File: EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    protected ResponseEntity<?> handleServerExceptionException(RuntimeException ex, WebRequest request) {
+        return handleExceptionInternal(ex,
+                Map.of("HTTP Status", "" + HttpStatus.INTERNAL_SERVER_ERROR, "response body", Map.of("message", ex.getMessage())),
+                new HttpHeaders(), HttpStatus.INTERNAL_SERVER_ERROR, request);
+    }
+
+    @ExceptionHandler(ApplicationException.class)
+    protected ResponseEntity<?> handleApplicationException(RuntimeException ex, WebRequest request) {
+        return handleExceptionInternal(ex,
+                Map.of("HTTP Status", "" + HttpStatus.INTERNAL_SERVER_ERROR, "response body", Map.of("message", "Went wrong")),
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java:37]: did , return Oops, something went wrong

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    public ResponseEntity<?> getById(@PathVariable long id) {
+        if (DataValidation.moreThenZero(id)) {
+
+            return ResponseEntity.ok(Map.of("gift certificate", giftCertificateService.getCertificateById(id)));
+        }
+        throw new ServerException("incorrect id=" + id);
+    }
+
+    @PatchMapping("/{id}")
+    public ResponseEntity<?> updateCertificate(@RequestBody GiftCertificate giftCertificate, @PathVariable long id) {
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/controller/GiftCertificateController.java:53]: done

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/model/GiftCertificate.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        return df.format(date);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof GiftCertificate)) return false;
+        GiftCertificate that = (GiftCertificate) o;
+        return Objects.equals(getId(), that.getId()) && Objects.equals(getName(), that.getName()) && Objects.equals(getTags(), that.getTags()) && Objects.equals(getDescription(), that.getDescription()) && Objects.equals(getPrice(), that.getPrice()) && Objects.equals(getDuration(), that.getDuration()) && Objects.equals(getCreateDate(), that.getCreateDate()) && Objects.equals(getLastUpdateDate(), that.getLastUpdateDate());
+    }
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/model/GiftCertificate.java:121]: done , I remake using commons3 


    public boolean equals(Object o) {
        if (this == o) return true;

        if (!(o instanceof GiftCertificate)) return false;

        GiftCertificate that = (GiftCertificate) o;

        return new EqualsBuilder().append(getId(), that.getId()).append(getName(), that.getName()).append(getTags(), that.getTags()).append(getDescription(), that.getDescription()).append(getPrice(), that.getPrice()).append(getDuration(), that.getDuration()).append(getCreateDate(), that.getCreateDate()).append(getLastUpdateDate(), that.getLastUpdateDate()).isEquals();
    }

    public int hashCode() {
        return new HashCodeBuilder(17, 37).append(getId()).append(getName()).append(getTags()).append(getDescription()).append(getPrice()).append(getDuration()).append(getCreateDate()).append(getLastUpdateDate()).toHashCode();
    }

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    @Override
+    public boolean createCertificate(GiftCertificate giftCertificate) {
+        return jdbcTemplate.update(SqlQuery.GiftCertificate.CREATE_CERTIFICATE, giftCertificate.getName(),
+                giftCertificate.getDescription(),
+                giftCertificate.getPrice(),
+                giftCertificate.getDuration(),
+                Instant.now(),
+                Instant.now()) == 1;
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:48]: done , remake Instant.now() to currentDateTime like const
1 remake to constant too

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    @Override
+    public long getIdByGiftCertificate(GiftCertificate giftCertificate) {
+        return jdbcTemplate.queryForObject(SqlQuery.GiftCertificate.GET_ID_BY_GIFT_CERTIFICATE,
+                new Object[]{giftCertificate.getName(), giftCertificate.getDescription(), giftCertificate.getPrice(), giftCertificate.getDuration()}
+                , Long.class);
+    }
+
+    @Override
+    public boolean isGiftCertificateExist(GiftCertificate giftCertificate) {
+        return Boolean.TRUE.equals(jdbcTemplate.queryForObject(SqlQuery.GiftCertificate.IS_GIFT_CERTIFICATE_EXIST, Boolean.class,
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:97]: already done , it is just what ide recommend me )

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        jdbcTemplate.update(SqlQuery.GiftCertificate.CREATE_GIFT_CERTIFICATE_TAG, giftCertificateId, tagId);
+    }
+
+    private void deleteGiftCertificateTag(long tagId, long giftCertificateId) {
+        jdbcTemplate.update(SqlQuery.GiftCertificate.DELETE_GIFT_CERTIFICATE_TAG, giftCertificateId, tagId);
+    }
+
+    @Override
+    public boolean createGiftCertificateTagRelationship(List<Long> tagsId, long giftCertificateId) {
+        for (long tagId : tagsId) {
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:111]: remake , thank u , while resolving this one , I find bug and fix it

File: EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/controller/GiftCertificateTagController.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        if (DataValidation.isStringValid(method)) {
+            if (DataValidation.isSortingTypeContain(method)) {
+                return ResponseEntity.ok(giftCertificateTagService.sortingAscDescByDate(method));
+            }
+            throw new ServerException("type should be only DESC/ASC without register");
+        }
+        throw new ServerException("Incorrect data");
+    }
+
+    @GetMapping(value = "sort-date-name/{method1}/{method2}")
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/controller/GiftCertificateTagController.java:53]: replace with sortingType

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        }
+        throw new ItemNotFoundException("There is no gift certificate with id= " + id);
+    }
+
+    @Transactional
+    public List<GiftCertificate> updateGiftCertificate(long id, List<Tag> tags, Optional<Map<String, String>> updatesMap) {
+
+        updatesMap.get().remove("tags");
+
+        if (giftCertificateRepository.updateGiftCertificate(id, updatesMap)) {
```
Comment (OlehHoliuk) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java:80]: Why do you pass Oprional `updatesMap` to repository if you already called `get` earlier?

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        if (giftCertificateRepository.deleteCertificate(id)) {
+            return true;
+        }
+        throw new ItemNotFoundException("There is no gift certificate with id= " + id);
+    }
+
+    @Transactional
+    public List<GiftCertificate> updateGiftCertificate(long id, List<Tag> tags, Optional<Map<String, String>> updatesMap) {
+
+        updatesMap.get().remove("tags");
```
Comment (OlehHoliuk) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java:78]: Calling get without checking if value is present will result in exception.

Comment (OlehHoliuk) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java:78]: Move `tags` to constants

File: EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/controller/GiftCertificateTagController.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        if (DataValidation.isStringValid(sortingType)) {
+            if (DataValidation.isSortingTypeContain(sortingType)) {
+                return ResponseEntity.ok(giftCertificateTagService.sortingAscDescByDate(sortingType));
+            }
+            throw new ServerException("type should be only DESC or ASC without register");
+        }
+        throw new ServerException("Incorrect data");
+    }
+
+    @GetMapping(value = "sort-date-name/{sortingType1}/{sortingType2}")
```
Comment (OlehHoliuk) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/controller/GiftCertificateTagController.java:53]: It is still unclear to me what does `sortingType1` and `sortingType2` mean. What is the difference between those 2?

File: EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/repository/GiftCertificateTagRepository.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (OlehHoliuk) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/repository/GiftCertificateTagRepository.java:12]: It is unclear to me, what it is sorting. I assume it returns sorted list of gift certificates sorted by date in ASC or DESC direction, based on the `method` variable. Am I right? 
If so then I'd suggest to rename method to `getGiftCertificatesSortedByDate(String sortDirection)` or smth similar

File: EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/exception/ServerException.java [hunk 17]
```diff
@@ -0,0 +1,7 @@
+package com.epam.esm.exceptionhandler.exception;
+
+public class ServerException extends ApplicationException {
```
Comment (OlehHoliuk) [inline EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/exception/ServerException.java:3]: What is the difference between `ApplicationException` and  `ServerException`? When do use one or another? Please provide examples

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/model/GiftCertificate.java [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+
+        if (!(o instanceof GiftCertificate)) return false;
+
+        GiftCertificate that = (GiftCertificate) o;
+
+        return new EqualsBuilder().append(getId(), that.getId()).append(getName(), that.getName()).append(getTags(), that.getTags()).append(getDescription(), that.getDescription()).append(getPrice(), that.getPrice()).append(getDuration(), that.getDuration()).append(getCreateDate(), that.getCreateDate()).append(getLastUpdateDate(), that.getLastUpdateDate()).isEquals();
```
Comment (OlehHoliuk) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertficate/model/GiftCertificate.java:125]: Reformat this mess

File: EEJT-0002/src/main/java/com/epam/esm/utils/validation/DataValidation.java [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+                map.put("tags", "some tags");
+            }
+            return Optional.of(map);
+        }
+        throw new ServerException("Something went wrong, check this fields " + answer);
+    }
+
+    public static boolean isStringValid(String obj) {
+        return obj != null && !StringUtils.isBlank(obj) && !StringUtils.isEmpty(obj) && !StringUtils.isNumeric(obj);
```
Comment (OlehHoliuk) [inline EEJT-0002/src/main/java/com/epam/esm/utils/validation/DataValidation.java:84]: You don't need to check for null explicitly, StringUtils methods already do it.
You can use StringUtils::isNotBlank, to get rid of inversion.
You don't need to check `isEmpty` after `isBlank`, it already checking this as well.

File: EEJT-0002/src/main/java/com/epam/esm/utils/validation/DataValidation.java [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    public static boolean isSortingTypeContain(String method) {
+        return List.of("ASC", "DESC").contains(method.toUpperCase(Locale.ROOT));
```
Comment (OlehHoliuk) [inline EEJT-0002/src/main/java/com/epam/esm/utils/validation/DataValidation.java:88]: Did you consider creating enum for sorting types?

**Review by OlehHoliuk — CHANGES_REQUESTED** [review]

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java (hunk 14 above)
Comment (OlehHoliuk) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java:80]: Optional should be used as return type, and not as entry parameter of the method

File: EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/repository/GiftCertificateTagRepository.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/repository/GiftCertificateTagRepository.java:12]: I did this , your variant is better , it seams to me that i make one method sorting norm but another one bad , ty

File: EEJT-0002/src/main/java/com/epam/esm/config/SwaggerConfig.java [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+import springfox.documentation.service.ApiInfo;
+import springfox.documentation.spi.DocumentationType;
+import springfox.documentation.spring.web.plugins.Docket;
+import springfox.documentation.swagger2.annotations.EnableSwagger2;
+
+import static springfox.documentation.service.ApiInfo.DEFAULT_CONTACT;
+
+@Configuration
+@EnableSwagger2
+public class SwaggerConfig {
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/config/SwaggerConfig.java:18]: oh , i considered that i delete this one , sorry mu bad

File: EEJT-0002/src/main/java/com/epam/esm/documentation/swagger.yml [hunk 22]
```diff
@@ -0,0 +1,573 @@
+swagger: '2.0'
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/documentation/swagger.yml:1]: resolved

File: EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.Map;
+
+@ControllerAdvice
+public class ResponseExceptionHandler extends ResponseEntityExceptionHandler {
+
+    @ExceptionHandler(ItemNotFoundException.class)
+    protected ResponseEntity<?> handleItemNotFoundException(ItemNotFoundException ex, WebRequest request) {
+        return handleExceptionInternal(ex,
+                Map.of("HTTP Status", "" + HttpStatus.NOT_FOUND, "response body", Map.of("message", ex.getMessage())),
+                new HttpHeaders(), HttpStatus.NOT_FOUND, request);
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java:24]: i made constants to String

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+public class GiftCertificateRepositoryImp implements GiftCertificateRepository {
+
+    private final JdbcTemplate jdbcTemplate;
+    private final String ID = "id";
+    private final String NAME = "name";
+    private final String DESCRIPTION = "description";
+    private final String PRICE = "price";
+    private final String DURATION = "duration";
+    private final String CREATE_DATE = "create_date";
+    private final String LAST_UPDATE_DATE = "last_update_date";
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:23]: did

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+    private final int successfullyDone = 1;
+
+    @Autowired
+    public GiftCertificateRepositoryImp(JdbcTemplate jdbcTemplate) {
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    private List<Tag> getAllTagsByCertificateId(long certificateId) {
+        return jdbcTemplate.query(SqlQuery.GiftCertificate.GET_ALL_TAGS_BY_CERTIFICATE_ID, new Long[]{certificateId}, (resultSet, i) ->
+                new Tag().setId(resultSet.getLong("id")).setName(resultSet.getString("name")));
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:34]: done

[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:34]: did

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+    @Override
+    public boolean deleteCertificate(long id) {
+        return jdbcTemplate.update(SqlQuery.GiftCertificate.DELETE_CERTIFICATE, id) == successfullyDone;
+    }
+
+    @Override
+    public boolean updateGiftCertificate(long id, Optional<Map<String, String>> updatesMap) {
+        List<Object> updatesParam = new ArrayList<>();
+
+        StringBuilder generatedQuery = new StringBuilder("UPDATE gift_certificate SET ");
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:83]: done

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 27]
```diff
@@ -0,0 +1,130 @@
...
+        updatesMap.get().forEach((key, value) -> {
+
+            generatedQuery.append(key).append("= ? ,");
+            updatesParam.add(value);
+        });
+        generatedQuery.append("last_update_date = ? WHERE id = ?");
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:89]: done

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+    private final JdbcTemplate jdbcTemplate;
+    private final String ID = "id";
+    private final String NAME = "name";
+    private final String DESCRIPTION = "description";
+    private final String PRICE = "price";
+    private final String DURATION = "duration";
+    private final String CREATE_DATE = "create_date";
+    private final String LAST_UPDATE_DATE = "last_update_date";
+    private final Instant currentDateTime = Instant.now();
+    private final int successfullyDone = 1;
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:25]: done

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+        updatesParam.addAll(List.of(currentDateTime, id));
+
+        return jdbcTemplate.update(generatedQuery.toString(), updatesParam.toArray()) == successfullyDone;
+    }
+
+    @Override
+    public long getIdByGiftCertificate(GiftCertificate giftCertificate) {
+        return jdbcTemplate.queryForObject(SqlQuery.GiftCertificate.GET_ID_BY_GIFT_CERTIFICATE,
+                new Object[]{giftCertificate.getName(), giftCertificate.getDescription(), giftCertificate.getPrice(), giftCertificate.getDuration()}
+                , Long.class);
```
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:100]: before that, I check if there is such a certificate in the database, in order to understand if such a name, etc., is in the database, when creating

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java (hunk 14 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java:80]: remake

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java (hunk 15 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java:78]: ty , did

Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java:78]: done

File: EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/controller/GiftCertificateTagController.java (hunk 16 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertificatetag/controller/GiftCertificateTagController.java:53]: remake

File: EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/exception/ServerException.java (hunk 17 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/exception/ServerException.java:3]: i remake this , it is just copypaste

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/model/GiftCertificate.java (hunk 18 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/model/GiftCertificate.java:125]: ok

File: EEJT-0002/src/main/java/com/epam/esm/utils/validation/DataValidation.java (hunk 20 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/utils/validation/DataValidation.java:88]: i do this now

File: EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (OlehHoliuk) [reply EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java]: Constants is good, but it didn't help with repetitiveness of the code. Extract common logic into separate method.

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+
+    @Override
+    public boolean createCertificate(GiftCertificate giftCertificate) {
+        return jdbcTemplate.update(SqlQuery.GiftCertificate.CREATE_CERTIFICATE,
+                giftCertificate.getName(),
+                giftCertificate.getDescription(),
+                giftCertificate.getPrice(),
+                giftCertificate.getDuration(),
+                currentDateTime.now(),
+                currentDateTime.now()) == DB_OPERATION_SUCCESSFUL_RESULT;
```
Comment (OlehHoliuk) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:60]: There is still no explanation what those `currentDateTime.now()` represent.
Create a variable (not class field) above with a proper name, assign `currentDateTime.now()` as value and use it here.

**Review by OlehHoliuk — CHANGES_REQUESTED** [review]

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java (hunk 30 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:60]: done

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java [hunk 31]
```diff
@@ -0,0 +1,140 @@
...
+    @Override
+    public boolean createCertificate(GiftCertificate giftCertificate) {
+        Instant currentTime = Instant.now();
+        return jdbcTemplate.update(SqlQuery.GiftCertificate.CREATE_CERTIFICATE,
+                giftCertificate.getName(),
+                giftCertificate.getDescription(),
+                giftCertificate.getPrice(),
+                giftCertificate.getDuration(),
+                currentTime,
+                currentTime) == DB_OPERATION_SUCCESSFUL_RESULT;
```
Comment (OlehHoliuk) [inline EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java:60]: You are getting closer. But passing currentTime twice still makes not a lot of sense.
What I wanterd is to rename first one to creationDate, and second one to updatedDate or smth similar. Which will expain with the naming the purpose of two `Instant.now()` usages

**Review by OlehHoliuk — CHANGES_REQUESTED** [review]

File: EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java (hunk 31 above)
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/giftcertficate/repository/GiftCertificateRepositoryImp.java]: done

File: EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (treep25) [reply EEJT-0002/src/main/java/com/epam/esm/exceptionhandler/handler/ResponseExceptionHandler.java]: done

**Review by OlehHoliuk — APPROVED** [review]

**Pull request merged**